prefect_gcp/cloud_run_job.py

```diff
@@ -0,0 +1,578 @@
+            status=execution["status"],
+            log_uri=execution["status"]["logUri"],
+        )
+
+
+class CloudRunJob(Infrastructure):
+    """Infrastructure block used to run GCP Cloud Run Jobs.
+
+    Project name information is provided by the Credentials object, and should always
+    be correct as long as the Credentials object is for the correct project.
+    """
+
+    _logo_url = "https://images.ctfassets.net/gm98wzqotmnx/4CD4wwbiIKPkZDt4U3TEuW/c112fe85653da054b6d5334ef662bec4/gcp.png?h=250"  # noqa
+    _block_type_name = "Cloud Run Job"
+
+    type: Literal["cloud-run-job"] = Field(
+        "cloud-run-job", description="The slug for this task type."
+    )
+    image: str = Field(
+        description=(
+            "The image to use for a new Cloud Run Job. This value must "
+            "refer to an image within either Google Container Registry "
+            "or Google Artifact Registry."
+        ),
+    )
+
+    region: str
+    credentials: GcpCredentials
+
+    # Job settings
+    cpu: Optional[int] = Field(
+        description=(
+            "The amount of compute allocated to the Cloud Run Job. The int must be valid "
+            "based on the rules specified at https://cloud.google.com/run/docs/configuring/cpu#setting-jobs ."
+        )
+    )
+    memory: Optional[str] = Field(
+        description=(
+            "The amount of compute allocated to the Cloud Run Job. The string must be valid "
+            "based on the rules specified at https://cloud.google.com/run/docs/configuring/memory-limits#setting-jobs ."
+        )
+    )
+    args: Optional[list[str]] = Field(
+        description=(
+            "Arguments to be passed to your Cloud Run Job's entrypoint command."
+        )
+    )
+    env: dict[str, str] = Field(
+        default_factory=dict,
+        description="Environment variables to be passed to your Cloud Run Job.",
+    )
+
+    # Cleanup behavior
+    keep_job_after_completion: Optional[bool] = True
+
+    # For private use
+    _job_name: str = None
+    _execution: Optional[Execution] = None
+    _project_id: str = None
+
+    @property
+    def project_id(self):
+        if self._project_id is None:
+            self._project_id = self.credentials.get_project_id()
+
+        return self._project_id
+
+    @property
+    def job_name(self):
+        """Create a unique and valid job name."""
+
+        if self._job_name is None:
+            components = self.image.split("/")
+            # gcr.io/<project_name>/repo/whatever
+            image_name = components[2]
+            modified_image_name = image_name.replace((":"), "-").replace(
+                ("."), "-"
+            )  # only alphanumeric and '-' allowed
+            name = f"{modified_image_name}-{uuid4()}"
+            if len(name) > 50:
+                self.logger.warning(
+                    "Job names must be 50 or fewer characters in length. Shortening "
+                    f"job name from {name} to {name[:50]}."
```

Comment:
https://peps.python.org/pep-3101/
```suggestion
                    f"job name from {name!r} to {name[:50]!r}."
```